netket/experimental/dynamics/_integrator.py

```diff
@@ -0,0 +1,436 @@
+        """
+        t_dtype = jnp.result_type(_dtype(t0), _dtype(dt))
+
+        return IntegratorState(
+            t=jnp.array(t0, dtype=t_dtype),
+            y=y0,
+            dt=jnp.array(dt, dtype=t_dtype),
+            solver=self._solver,
+            last_norm=0.0 if self.use_adaptive else None,
+            last_scaled_error=0.0 if self.use_adaptive else None,
+            flags=IntegratorFlags(0),
+        )
+
+    def step(self, max_dt: float = None) -> bool:
+        """
+        Performs one full step by :code:`min(self.dt, max_dt)`.
+
+        Returns:
+            A boolean indicating whether the step was successful or
+            was rejected by the step controller and should be retried.
+
+            Note that the step size can be adjusted by the step controller
+            in both cases, so the integrator state will have changed
+            even after a rejected step.
+        """
+        self._state = self._do_step(
+            solver=self._solver,
+            f=self.f,
+            state=self._state,
+            atol=self.atol,
+            rtol=self.rtol,
+            norm_fn=self.norm,
+            max_dt=max_dt,
+            dt_limits=self.dt_limits,
+        )
+        return self._state.accepted
+
+    @property
+    def t(self) -> float:
+        """The actual time."""
+        return self._state.t.value
+
+    @property
+    def y(self) -> PyTree:
+        """The actual state."""
+        return self._state.y
+
+    @property
+    def dt(self) -> float:
+        """The actual time-step size."""
+        return self._state.dt
+
+    @property
+    def solver(self) -> "AbstractSolver":
+        """The ODE solver."""
+        return self._solver
+
+    def _get_integrator_flags(self, intersect=IntegratorFlags.NONE) -> IntegratorFlags:
+        r"""Returns the currently set flags of the integrator, intersected with `intersect`."""
+        # _state.flags is turned into an int-valued DeviceArray by JAX,
+        # so we convert it back.
+        return IntegratorFlags(int(self._state.flags) & intersect)
+
+    @property
+    def errors(self) -> IntegratorFlags:
+        r"""Returns the currently set error flags of the integrator."""
+        return self._get_integrator_flags(IntegratorFlags.ERROR_FLAGS)
+
+    @property
+    def warnings(self) -> IntegratorFlags:
+        r"""Returns the currently set warning flags of the integrator."""
+        return self._get_integrator_flags(IntegratorFlags.WARNINGS_FLAGS)
+
+    def __repr__(self) -> str:
+        return "{}(solver={}, state={}, adaptive={}{})".format(
+            "Integrator",
+            self.solver,
+            self._state,
+            self.use_adaptive,
+            (
+                f", norm={self.norm}, atol={self.atol}, rtol={self.rtol}, dt_limits={self.dt_limits}"
+                if self.use_adaptive
+                else ""
+            ),
+        )
+
+
+@partial(maybe_jax_jit, static_argnames=["f"])
+def general_time_step_fixed(
```

Comment:
is there a reason for keeping those functions as standalone functions instead of making them methods of the Integrator class? Like `_step_fixed` and `_step_adaptive`?

I think that now with the new interface we should not need to use them elsewhere (correct me If I am wrong).
If I'm wrong, let's keep them here, but otherwise it would be nice, for consistency and easier experimentation to put everything in the integrator. Also makes it easier to subclass integrator but keep them in there.

Reply:
I don't think there is a reason anymore (not sure why it was like this even before... maybe to be able to use it to re-define ?). But, indeed, it should work internally. Let's change it